notes/plutus-report/costing.tex

```diff
@@ -0,0 +1,339 @@
+\item The initial Space cost of the AST.
+\item The initial Time cost for starting the machine (the intercept of the linear model for the machine steps).
+\item The Time and Space costs for each machine step that is taken (as determined by the coefficients of the linear model for the machine steps).
+\item The Time and Space costs for each builtin call (as determined by the linear models for each builtin operation).
+\end{itemize}
+
+\subsection{Building the pricing model}
+The pricing model is out of our control - it is the ledger's responsibility, and as we discussed earlier, the process of choosing the model is likely to be complex and influenced by many non-technical factors.
+However, anyone making a decision about how to set prices does need at least some input from us, because prices relate to real costs (real CPU seconds, real memory usage), and so we need to know how our abstract units correspond (in reality, at the current time) to real units.
+
+Fortunately, we have this information easily to hand, since we are targeting specific relationships between abstract units and real units when we build our models.
+So we can assume that those relationships hold (e.g. 1 Time = 1 microsecond).
+
+\subsection{Security}
+The top priority of the model is to ensure the security of the ledger against attacks, particularly DoS attacks, but also economic attacks that sap profit from the system.
+
+\paragraph{Basic structure of an attack}
+The basic kind of attack we are worried about is where we have set one of our cost parameters too low, effectively under-costing a particular operation.
+For example, maybe addition is too cheap, or a certain kind of machine step.
+
+An attacker could then construct a synthetic program that disproportionately uses the under-costed operation.
+Such a program would then be more expensive to execute in reality than the model predicts, allowing the attacker to force node operators to do more work than they are paying for.
+
+How dangerous this is depends on how badly we under-estimate the parameter.
+If we under-estimate it by only 10\%, then an attacker is only getting a 10\% ``discount'' on computation: probably not enough to make it worthwhile to mount an attack.
+If we under-estimate it by an order of magnitude or two, then we could be in trouble.
+
+Even then, a malicious program may have to be quite large, and attackers may run up against transaction size limits.
+
+\paragraph{Simple prevention approaches}
+Our attack prevention approach is fairly simple.
+We try and make the cost model give an over-estimate of reality, and then we try to make the pricing model over-price the resources.
+
+In order to keep this working, we need to:
+\begin{enumerate}
+\item Ensure that we take great care when updating the cost model, especially if we make things cheaper (say, on the basis of the evaluator getting faster)
+\item Ensure that we adjust prices as appropriate when the cost of hardware changes.
+\end{enumerate}
+
```

Comment:
Maybe we could refer (probably elsewhere) to that paper on "Empirically Analyzing Ethereum's Gas Mechanism" (https://arxiv.org/abs/1905.00553) and say something about how what we're doing compares.  Maybe that's not necessary in this particular document though.

Reply:
Yeah, it would be nice to say something about that, although I don't have much to say at the moment.

Reply:
I had a quick look at it a while ago and I think one of the main things that they pointed out was that there was an instruction which performed disk access that was under-costed. At least we're not doing that during evaluation.